torch/csrc/jit/codegen/cuda/lower_scalar_hoist.cpp

```diff
@@ -245,26 +268,45 @@ class CommonIndexInserter : private kir::ExprMutator {
     if (loop != nullptr) {
       scope = &loop->body();
     }
-    for (auto index : common_index_map_.getHoistedIndices(loop)) {
-      // Make the type of the hoisted index be the index type of the
+    const auto& exprs = (scope == nullptr ? exprs_ : scope->exprs());
+    int64_t alloc_point = -1;
+    Expr* insert_ref = nullptr;
+    for (auto value : common_scalar_map_.getHoistedScalars(loop)) {
+      auto existing_alloc_info = findAllocPointFromDataDependency(exprs, value);
+      // If this value has already been fully computed, then don't insert
```

Comment:
For example, in `NVFuserTest.FusionStandaloneARange_CUDA`, without this check, we will get generated kernel:
```CUDA
__global__ void kernel1(int64_t i2, int64_t i1, int64_t i0, Tensor<float, 1> T0) {
  int i58;
  i58 = (((nvfuser_index_t)blockIdx.x) * 128) + ((nvfuser_index_t)threadIdx.x);
  double d8;
  d8 = (double)(i2);
  double d7;
  d7 = (double)(i1);
  double d6;
  d6 = (double)(i0);
  double d6;
  d6 = (double)(i0);
  double d8;
  d8 = (double)(i2);
  if ((i58 < ((int64_t)((ceilDiv((abs((d7 - d6))), (abs(d8)))))))) {
    float T4[1];
    T4[0] = (((float)(d6)) + (((float)(d8)) * ((float)(i58))));
    T0[i58]
       = T4[0];
  }
}
```
where `d6` and `d8` is allocated twice. I think this is because in the definition of `arange` in `arith.cpp`, we have
```C++
TensorView* arange(Val* start, Val* end, Val* step, DataType dtype) {
  if (isIntegralType(dtype)) {
    start = castOp(DataType::Int, start);
    end = castOp(DataType::Int, end);
    step = castOp(DataType::Int, step);
  } else if (isFloatingPointType(dtype)) {
    start = castOp(DataType::Double, start);
    end = castOp(DataType::Double, end);
    step = castOp(DataType::Double, step);
  }
  // Make sure no negative value is passed to ceilDiv as the device
  // implementation of ceilDiv assumes positive inputs
  auto size = castOp(DataType::Int, ceilDiv(abs(sub(end, start)), abs(step)));
  auto out = TensorViewBuilder()
                 .ndims(1)
                 .dtype(dtype)
                 .contiguity({true})
                 .shape({size})
                 .build();
  IrBuilder::create<ARangeOp>(out, start, end, step, dtype);
  return out;
}
```
where cast op is used to get the correct type to calculate the ID extent.

Reply:
I see. When a value is an output of an expression, does it also mean there exists an allocation of the value? Actually, I don't remember where we create an allocation node for those scalar values like `d8`. `insertAllocations` seems to only do so for tensors.

Reply:
I think they are inserted in `lower_loops.cpp`:
```C++
void LoopNestGenerator::handle(Expr* expr) {
  // Check if it's a tensor view expression we need to place in the loop nest
  // structure
  if (!ir_utils::isTvOp(expr)) {
    // Close all the loops, scalar operations cannot be inside for loops based
    // on expr sorting.
    while (!for_loops_.empty()) {
      closeFor();
    }
    pushFront(expr);

    for (auto out : expr->outputs()) {
      TORCH_INTERNAL_ASSERT(
          out->getValType().value() == ValType::Scalar,
          "Unrecognized output type found in expr ",
          expr,
          " cannot lower ",
          out->getValType().value());

      pushFront(IrBuilder::create<kir::Allocate>(
          out, MemoryType::Local, GpuLower::current()->kernel()->oneVal()));
    }
    return;
  }
...
```

Reply:
Oh, thanks. Didn't expect it's done there.

So, any scalar val created after this pass doesn't have an allocation. Doesn't that mean we may need to insert an allocation of a hoisted val?

Reply:
> Doesn't that mean we may need to insert an allocation of a hoisted val?

Yes, and isn't that what `allocateCommonScalars` is doing?

Reply:
Hmm, but there's this just below:

```
if (existing_alloc_info.second) {
        continue;
 }
```

So, it skips the allocation when it's already computed, but does that guarantee it's allocated?

Reply:
Hmm, I am not sure if I understand your question. So, if `existing_alloc_info.second` is true, then there is an expr already in `exprs` that compute `value`, which according to `void LoopNestGenerator::handle(Expr* expr)`, `value` should have already been allocated, so we should not allocate it again in this function. Does this make sense? Hmm, are you saying that, there might be a case where there is a `value1` is already allocated, and `value` satisfies `value->sameAs(value1)` but not `value == value1`, and `value` is not allocated yet?

Reply:
Ah, you're right. This is only looking at expressions at a scope level, so there must be allocations. Sorry for my confusion.